### PR TITLE
Include a self-signed certificate in the influxdb image

### LINKED ISSUE #23: Add option to enable HTTPS
I'm hunting a bug in a Node client library that might be triggered by HTTPS.

### PR DESCRIPTION
**Pull request by jsternberg**
Generate a self-signed certificate and include it in InfluxDB by
default. This way, someone could start the HTTPS version of InfluxDB for
testing without having to generate their own certificate and then
mounting a filesystem or creating their own image.

Since HTTPS is disabled by default, we aren't shipping an image that
with an insecure certificate. A person who wants to ship with a
secure certificate is going to need to replace the certificate anyway.

HTTPS can be enabled using environment variables so it can use this
self-signed certificate.

Fixes #23.

### CODE REVIEW

File: influxdb/1.0/genssl.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
Comment (e-dard) [inline influxdb/1.0/genssl.sh]: This isn't used or executed by anything as far as I can tell?

**Comment by jsternberg** [comment]
My rationale was that the self-signed certificate isn't going to be secure and it's not going to be production ready. I generated it outside of the build image because I didn't want to spend a docker command having to generate what is essentially boilerplate and there didn't seem to be a lot of reason to keep the private key secret. If we generate it in `build`, it's just as good as exposing the private key anyway because somebody can just run this: `docker run --rm influxdb cat /path/to/file` and now they have the "private" key, so it's not really private.

I could modify the entrypoint script to generate the ssl certificate, but I didn't want to slow down startup of an image that's not using SSL.

**Comment by e-dard** [comment]
>  If we generate it in build, it's just as good as exposing the private key anyway because somebody can just run this: docker run --rm influxdb cat /path/to/file and now they have the "private" key, so it's not really private.

But the key is only exposed to an operator who has access to the server/system in question isn't it? That seems totally fine to me. Or are you saying that anyone who build the image would generate the exact same key/cert? 

I agree that we don't need to generate the certs during the image build and we could probably get away with bundling pre-determined certs but something is just niggling in the back of my head. Mainly over two scenarios:

 - Someone will not understand that this is for testing, and blog/tweet/blah that we're shipping an insecure setup;
 - Someone will switch on HTTPS but not replace the certificates and therefore be running a completely insecure setup.

> I could modify the entrypoint script to generate the ssl certificate, but I didn't want to slow down startup of an image that's not using SSL.

Will it add a lot of time to building the image? Minutes? Could you maybe use [ARG](https://docs.docker.com/engine/reference/builder/#/arg) to create build-time arguments for generating self-signed certs if the slowness of installing `openssl` and generating a cert is a significant issue?

**Comment by nathanielc** [comment]
> Someone will switch on HTTPS but not replace the certificates and therefore be running a completely insecure setup.

This is the scenario that concerns me, while it may not be frequent, as our user base grows a  larger and larger number of users will make this mistake. At some point it may become advantageous for an attacker to exploit the fact that there are tons of infuxdb server running in the wild with the same private key. 

By generating it in the build process at least the key changes frequently. Or better yet use a build ARG like @e-dard  suggested.

**Comment by jsternberg** [comment]
> But the key is only exposed to an operator who has access to the server/system in question isn't it? That seems totally fine to me. Or are you saying that anyone who build the image would generate the exact same key/cert?

Nope. Server private keys have to be on the server instance running. As can be seen, `genssh.sh` doesn't actually need any input. The private key has to be there when the server starts up. Traditionally when deploying SSL's in a real secure environment, you use some other method to transfer the SSL certificate right before the server starts rather than say, baking it into an image. Baking it into an image would be insecure. So I'm mostly saying that the private key has to end up in the image somehow. Either it's done through mounting a volume or it's baked into the image. The latter will always be insecure.

> Someone will not understand that this is for testing, and blog/tweet/blah that we're shipping an insecure setup;

I'm afraid of that too, but to be fair, I'll write in the documentation that we ship an insecure certificate and HTTPS libraries will complain almost immediately if you try and connect to the server. You have to explicitly turn off checking the SSL certificate's chain or add the chain to your trusted certificates for those libraries to work. If they're going through and doing that, then hopefully they know what they are doing.

> Someone will switch on HTTPS but not replace the certificates and therefore be running a completely insecure setup.

See above. HTTP libraries just won't connect to the server. This is also not the default configuration for the server.

And to be clear, there is no build argument ever in existence that will make this secure. It cannot be done at build time.

**Comment by e-dard** [comment]
@jsternberg I guess where you're coming from is that using self-signed certificates at all is insecure, which is fair enough. Where I'm coming from though is that there are going to be users who want to use self-signed certificates (for whatever reason) and they're going to screw up the setup and end up using the certs in this PR.

All they need to do to expose their data is run `influx -k -ssl`. At least if different  certs/keys are generated every time the image is built it won't be trivial for someone to decrypt the user's traffic.

I realise that this is all moot if we define secure as trustworthy as well as encrypted, but I'm making the assumption that some users will be comfortable trusting the machine/cert they're connecting to and just want to encrypt transport.

Does that make sense?

**Comment by jsternberg** [comment]
Yes, it makes sense, but I think at that point we might want to close this and make a feature request for InfluxDB. My rationale is:

1. We do not want to ship any self-signed certificate. This both is insecure because of verifying the host and also because the private key is exposed. Users may not understand this and may think `INFLUXDB_HTTP_HTTPS_ENABLED=true` makes it secure. It does not.
2. We do not want to incur a penalty for generating a self-signed certificate every time that the image starts even when HTTPS is disabled.
3. Figuring out if we need to generate the SSL certificate by parsing command line options, reading a TOML file, or trying to look at environment variables (all of which would need to be done) may be too difficult.

It might be good to have an disabled option in InfluxDB for generating a self-signed certificate. We can generate and sign it ourselves on startup and use that if they set the option. Likely more secure, but then we have to ask the question of "Is inside InfluxDB and appropriate place to put this functionality or should we hack it into `entrypoint.sh`?

**Comment by jsternberg** [comment]
I'm going to close this. I think there are too many complications with this that's it not worth doing and it's fairly easy to just do it yourself.

**Pull request closed without merging**